common/djangoapps/util/log_sensitive.py

```diff
@@ -0,0 +1,175 @@
+"""
+Utilities for logging sensitive debug information such as authentication tokens.
+
+Usage:
+
+1. Read the warning in cli_gen_keys
+2. Generate keys using ``python3 -m common.djangoapps.util.log_sensitive gen-keys``
+   and follow the instructions it prints out
+3. When logging sensitive information, use like so::
+
+     logger.info(
+         "Received invalid auth token %s in Authorization header",
+         encrypt_for_log(token, settings.AUTH_ERRORS_DEBUG_KEYS)
+     )
+
+   This will log a message like::
+
+     Received invalid auth token ZXIzljFT7zTTuXyIB/jy4XTixL34EhI6pY0Gw4mxfFo=|IYS4JDE0S9wypALb5HfoeW9YB/CnSx35/PkspsecZyzTP1Zr8B/W+16Gfe+pRZhjBitQ66XTZk9YHwKfznQ1KA== in Authorization header
+
+4. If you need to decrypt one of these messages, save the encrypted portion
+   to file, retrieve the securely held private key, and run::
+
+     python3 -m common.djangoapps.util.log_sensitive decrypt --message-file ... --private-key-file ...
+
+   If possible, use bash process indirection to keep the private key from
+   touching disk unencrypted, e.g. ``--private-key-file <(gpg2 --decrypt auth-logging-key.enc)``
+"""
+
+from base64 import b64decode, b64encode
+
+import click
+from nacl.public import Box, PrivateKey, PublicKey
+
+# Background:
+#
+# The NaCl "Box" construction provides asymmetric encryption, allowing
+# the sender to encrypt something for a recipient without having a
+# shared secret. This encryption is authenticated, meaning that the
+# recipient verifies that the message matches the sender's public key
+# (proof of sender). But it's also *repudiable* authentication; the
+# design allows both the sender and the receiver to read (or have
+# created!) the encrypted message, so the receipient can't prove to
+# anyone *else* that the sender was the author.
+#
+# Why we use ephemeral sender keys:
+#
+# The Box is normally an ideal construction to use for
+# communications. However, we don't want the logger to be able to read
+# the messages it writes, especially not messages from a different
+# server instance or from days or weeks ago. Only developers (or
+# others) in possession of the recipient keypair should be able to
+# read it, not anyone who compromises a server at some later
+# date. Luckily, we also don't care about authenticating the logged
+# messages as truly being from the server! The solution is for each
+# server to create a fresh public/private keypair at startup and to
+# include a copy of the public key in any encrypted logs it writes.
+
+
+# Generate an ephemeral private key for the logger to use during this
+# logging session.
+logger_private_key = PrivateKey.generate()
+
+
+def encrypt_for_log(message, reader_public_key_b64):
+    """
+    Encrypt a message so that it can be logged but only read by someone
+    possessing the given public key. The public key is provided in base64.
```

Comment:
I'm not clear. Should this read something like the following? If not, can you help me understand?
```suggestion
    Encrypt a message so that it can be logged with the public key, but
    only read by someone possessing the matching private key. The
    public key is provided in base64.
```

Reply:
Oh, yeah, just miswrote that.